rfc/004-priority-based-mempool.md

```diff
@@ -0,0 +1,205 @@
+
+## Proposal
+
+Add a new field `priority` to the ABCI `ResponseCheckTx` message:
+
+```proto
+message ResponseCheckTx {
+  uint32         code       = 1;
+  bytes          data       = 2;
+  string         log        = 3;  // nondeterministic
+  string         info       = 4;  // nondeterministic
+  int64          gas_wanted = 5 [json_name = "gas_wanted"];
+  int64          gas_used   = 6 [json_name = "gas_used"];
+  repeated Event events     = 7
+      [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string codespace = 8;
+  uint64 priority  = 9;
+}
+```
+
+Then ABCI developers can add application-specific user-defined priorities for the transactions. This addition is naturally backward-compatible because by default all requests will be returned with 0 priority and on mempool side the process will fallback to FIFO.
+
+The next step is to make current mempool implementation priority-aware. At some point switching from the CList to other data structures is necessary (e.g. self-balancing AVL tree or max-heap). [QuarkChain](https://github.com/tendermint/spec/pull/154) suggests using a Red-Black tree.
+
+When choosing between different BST, AVL was picked because it’s good for in-memory usage (comparing to BTree, which is often used in databases where persistence is needed) and fast retrieval (comparing to Red-Black tree, which is better for
+insertion/deletion due to relaxed rebalancing).
+
+Note that it's also needed to run tests to see if a map (similar to existing sync.Map) is needed to provide amortised O(1) access to individual elements (for retrieving/recording senders).
+
+We can have 3 phases to gradually implement this feature to make as least breaking changes as possible.
+
+### Phase 0: Keep existing CList-based mempool
+
+Goal:
+
+1. Zero cost on processing "legacy" transactions (i.e. those have 0 priority)
+2. When creating a block, transactions with higher priorities should be included earlier
+
+The first phase will be experimental and only achieves the desired functionality *with potential performance degradation*. The main purpose is to design the code flow and underlying APIs accordingly, without dramatic code changes on existing data structures / algorithms.
+
+A simple approach we can take is the following:
+
+1. When reading `CheckTx` responses, mempool keeps a single counter called `maxPriority`. If all transactions have 0 priorities, the counter should simply have the same 0 value
+2. When `Mempool.ReapMaxBytesMaxGas` is called by consensus engine, mempool would check its `maxPriority` value:
+    - if it's zero, it should follow existing logic to return eligible transactions in FIFO style
+    - if not, should return eligible transactions ranked by their priorities (e.g. `O(nlogn)` sort). Then during commit, update mempool's `maxPriority` accordingly since transactions will be removed from mempool
+
+### Phase 1: Abstract away existing mempool interface
```

Comment:
Seems like a great idea and should move forward independently of the rest of this RFC - probably best to move this part to an ADR in Tendermint directly (it doesn't break protocols or APIs so no need for spec-based RFC process, just changes to Tendermint Go I think) and work on it there while we figure out how to actually support the protocol changes for prioritization

Reply:
sounds good. I'll draft an ADR covering details the refactoring.

Reply:
update: ADR proposed in https://github.com/tendermint/tendermint/pull/5451